### DISCUSSION TITLE
ValueError: not find model.pdmodel or inference.pdmodel in output/v3_en_mobile/best_model

### DISCUSSION THREAD

**Discussion by kishore-work-hard** [PaddlePaddle/PaddleOCR#13085] [Q&A]
I found answers in Chinese but I don't know Chinese

 But i trained a model successfully. and in `output/v3_en_mobile/best_model` path folder i have `model.pdopt` and `model.pdparams`. 

after training my recognition model 
i used the command to test my script - 

`python tools/infer/predict_rec.py --image_dir="img/a (6).jpg" --rec_model_dir="output/v3_en_mobile/best_model" --rec_image_shape="3, 32, 100" --rec_char_dict_path="ppocr/utils/en_dict.txt"`

and got this error - 

```
Traceback (most recent call last):
  File "/home/elvicto/.ghostshell/Test/OCR/PaddleOCR/tools/infer/predict_rec.py", line 756, in <module>
    main(utility.parse_args())
  File "/home/elvicto/.ghostshell/Test/OCR/PaddleOCR/tools/infer/predict_rec.py", line 718, in main
    text_recognizer = TextRecognizer(args)
  File "/home/elvicto/.ghostshell/Test/OCR/PaddleOCR/tools/infer/predict_rec.py", line 149, in __init__
    ) = utility.create_predictor(args, "rec", logger)
  File "/home/elvicto/.ghostshell/Test/OCR/PaddleOCR/tools/infer/utility.py", line 214, in create_predictor
    raise ValueError(
ValueError: not find model.pdmodel or inference.pdmodel in output/v3_en_mobile/best_model
```

 dict used - `ppocr/utils/en_dict.txt`

after getting this error I tried the below code. 
and using the code - I figured it that this script downloads file automatically from this URL
 `https://paddleocr.bj.bcebos.com/PP-OCRv4/chinese/ch_PP-OCRv4_rec_infer.tar`

```
import os
import cv2
import numpy as np
import paddle
from paddleocr import PaddleOCR

# Define paths
model_dir = "D:/DATA/OCR_MODEL/best_model"
image_path = "../images/a (2).jpg"

# Load custom OCR model
ocr = PaddleOCR(
    rec_model_dir=model_dir,
    rec_char_dict_path=os.path.join(model_dir, "dict.txt"),  # Update with actual dictionary path if needed
    use_space_char=False
)

# Load image
img = cv2.imread(image_path)
if img is None:
    raise ValueError(f"Image at path {image_path} could not be loaded.")

# Perform OCR
result = ocr.ocr(img, det=False)

print(result)
```

But I want english model. so i replaced it with en and english
should i be downloading the file from this - 
 `https://paddleocr.bj.bcebos.com/PP-OCRv4/english/en_PP-OCRv4_rec_infer.tar`

what i mean is... 
after custom training my rec model. should I be using this files from en_PP-OCRv4_rec_infer.tar after downloading it.?
 
Or

should this file be used for pretrained model and should not be used by custom trained model? 
please explain what's happening. and what is this` inference.pdmodel` and `inference.pdiparams` ?

**Comment by SWHL** [comment]
You can refer this [doc](https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/recognition_en.md#32-test)

`inference.pdmodel` and `inference.pdiparams` are converted from the `model.pdopt` and `model.params`.

**Reply by kishore-work-hard** [reply]
> You can refer this [doc](https://github.com/PaddlePaddle/PaddleOCR/blob/main/doc/doc_en/recognition_en.md#32-test)
> 
> `inference.pdmodel` and `inference.pdiparams` are converted from the `model.pdopt` and `model.params`.

SO YOU ARE TELLING ME NOT TO DOWNLOAD THE MISSING FILES FILE FROM THE URL. ?
`https://paddleocr.bj.bcebos.com/PP-OCRv4/english/en_PP-OCRv4_rec_infer.tar` 

I didn't get the file `inference.pdmodel` and `inference.pdiparams` after training. 

can you tell me how to get it?

**Reply by pcycccccc** [reply]
你是拿训练完的模型文件进行测试，所以出现了这样的问题。
正确做法应该是：将你的训练完的模型转为推理模型，示例指令为：`python tools/export_model.py -c "./configs/rec/PP-OCRv3/en_PP-OCRv3_rec.yml" -o Global.pretrained_model="./output/v3_en_rec/best_accuracy.pdparams" Global.save_inference_dir="./inference
_model/rec/"`
你需要根据你的实际情况，更换一下路径。然后再利用你之前的测试脚本指令：python tools/infer/predict_rec.py来指向你的推理模型

**Comment by pcycccccc** [comment]
@kishore-work-hard 
You encountered issues with finding model.pdmodel or inference.pdmodel because you are testing with the model file obtained after training.

If you want to test your trained model, you need to follow these steps:

- 1、 Convert the trained model into an inference model. The example command is:
‘python tools/export_model.py -c "./configs/rec/PP-OCRv3/en_PP-OCRv3_rec.yml" -o Global.pretrained_model="./output/v3_en_rec/best_accuracy.pdparams" Global.save_inference_dir="./inference_model/rec/" ’
You need to adjust the paths according to your situation. It will automatically generate model files such as model.pdmodel or inference.pdmodel in the specified directory.

- 2、Use your previous test script command:
‘python tools/infer/predict_rec.py’
Make sure to set rec_model_dir to point to the inference model generated in Step One.

By following these two steps, you will resolve the issue you encountered.

**Reply by kishore-work-hard** [reply]
> * python tools/export_model.py -c "./configs/rec/PP-OCRv3/en_PP-OCRv3_rec.yml" -o Global.pretrained_model="./output/v3_en_rec/best_accuracy.pdparams" Global.save_inference_dir="./inference_model/rec/"

thanks for the English answer. I will try it out.

**Reply by Mishatopkek** [reply]
I don't get the second step. I managed to generate the same data that I can download from the webpage, but I can't run predict_rec.py. I just get random errors. I don't know what argument I should give

**Comment by pcycccccc** [comment]
@kishore-work-hard 
If you want to test the text recognition effect of your trained model after completing the training, you only need to use the model file obtained from your training. This does not involve the inference model files provided by the official source, such as ch_PP-OCRv4_rec_infer.tar or other inference model files.

I guess the script automatically downloads URL files because it cannot find your model.pdmodel or inference.pdmodel files during inference testing, so it downloads the inference model from the default URL.

**Reply by kishore-work-hard** [reply]
> @kishore-work-hard If you want to test the text recognition effect of your trained model after completing the training, you only need to use the model file obtained from your training. This does not involve the inference model files provided by the official source, such as ch_PP-OCRv4_rec_infer.tar or other inference model files.
> 
> I guess the script automatically downloads URL files because it cannot find your model.pdmodel or inference.pdmodel files during inference testing, so it downloads the inference model from the default URL.

I will followed ur instructions and create inference.pdmodel. using `tools/export_model.py`. 
It worked. 
results are still bad. 
how many images are needed to train an rec model for good accuracy. currently i used 10k cropped number plate images without using data argumentation. any insight would be helpfull

**Reply by pcycccccc** [reply]
@kishore-work-hard Did you use crop_image for both training and testing? Additionally, does your --rec_char_dict_path point to your dictionary file?

**Reply by kishore-work-hard** [reply]
> @kishore-work-hard Did you use crop_image for both training and testing? Additionally, does your --rec_char_dict_path point to your dictionary file?

dictionary path are all correct. i am getting one or two letters recognised out of 10 letters. 
both my training and testing data contains custom dataset from real life cropped number plates from the images of cars and bikes taken from ip cams.

i didnt train paddle's text det model. i disabled it and enabled only rec model when doing inference 
i only trained paddle's text rec model
(set rec=True, Det=False)

sample camera image(without cropping) -

![a (1)](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/f6cf360d-92c2-41b2-b0af-65851f1d9171)

sample dataset images (after cropping the number plate using yolo)- 

![1v_word_70_172 20 20 207_20240530092249231_KL01R7124_Tricycle](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/bfad8ea6-0dd3-47ba-aa3f-4ab21917580b)
![a (1)](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/9bcd182a-9035-4f63-836c-eb2e519692b3)
![a (2)](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/fd8905d2-3d57-425c-942d-e6250bd2af61)
![a (3)](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/0a5a46ab-6bd3-4c7f-983a-513702b55d35)
![a (7)](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/7e58b702-22ca-40e4-b85f-673837d78bb8)
![b](https://github.com/PaddlePaddle/PaddleOCR/assets/93419928/f7055a1e-938f-41bd-81b3-de1a50aa4c21)

These Cropped Number Plates are the datasets used as training and testing dataset.
